### PR TITLE
function: capture metric appearance for xrate/xincrease/xdelta

### PR DESCRIPTION
**Pull request by GiedriusS**
Current check for zero value injection doesn't make sense - we need to check it for since the beginning of when the metric appears.

### CODE REVIEW

**Comment by GiedriusS** [comment]
@mklocke-shopify what do you think about this as you've worked on the original implementation in this engine?

File: execution/scan/matrix_selector.go [hunk 1]
```diff
@@ -32,12 +32,13 @@ type matrixScanner struct {
...
-	storage  engstore.SeriesSelector
-	call     function.FunctionCall
-	scanners []matrixScanner
-	series   []labels.Labels
-	once     sync.Once
+	funcExpr       *parser.Call
+	storage        engstore.SeriesSelector
+	call           function.FunctionCall
+	scanners       []matrixScanner
+	metricAppeared []int64
```
Comment (fpetkovski) [inline execution/scan/matrix_selector.go:39]: Should we put this in the scanner itself? Also does it denote the time when the metric first appeared?

File: execution/scan/matrix_selector.go [hunk 2]
```diff
@@ -155,13 +156,14 @@ func (o *matrixSelector) Next(ctx context.Context) ([]model.StepVector, error) {
...
 			// https://github.com/thanos-community/promql-engine/issues/39
-			result := o.call(function.FunctionArgs{
-				Labels:      series.labels,
-				Samples:     rangeSamples,
-				StepTime:    seriesTs,
-				SelectRange: o.selectRange,
-				Offset:      o.offset,
-			})
+			fa := function.NewFunctionsArgs()
+			fa.Labels = series.labels
```
Comment (fpetkovski) [inline execution/scan/matrix_selector.go:160]: Does it make sense to have a constructor with all parameters so we don't accidentally omit something?

**Review by mklocke-shopify — APPROVED** [review]
This looks good! Thanks a lot for adding this, it makes sense.

**Comment by GiedriusS** [comment]
Thx for the review @fpetkovski, I have refactored the code to be more Go-like:

- Instead of `-1` to indicate the absence of a value, let's use a pointer to int64; this makes the zero value of FunctionArgs and others make sense;
- Moved `metricAppeared` into scanners;
- Appended `ts` suffix everywhere so that it would be clear that it stores a timestamp.

File: execution/scan/matrix_selector.go [hunk 3]
```diff
@@ -366,10 +368,20 @@ loop:
 		case chunkenc.ValNone:
 			break loop
 		case chunkenc.ValHistogram, chunkenc.ValFloatHistogram:
+			t, h := buf.AtHistogram()
```
Comment (fpetkovski) [inline execution/scan/matrix_selector.go:371]: Hmm, we call `AtFloatHistogram` bellow, so I wonder why we need to call `AtHistogram` here. We also seem to now append to the output before we check for staleness, so we might end up picking up stale samples. Is this the correct behavior?

**Review by fpetkovski — APPROVED** [review]

File: execution/scan/matrix_selector.go (hunk 3 above)
Comment (GiedriusS) [reply execution/scan/matrix_selector.go]: You are correct, this was an artifact of rebasing. I somehow missed this part. Thank you for a diligent review :bow:

**Pull request merged**